lucene/codecs/src/java/org/apache/lucene/codecs/uniformsplit/IntersectBlockReader.java

```diff
@@ -18,260 +18,337 @@
     super(dictionaryBrowserSupplier, blockInput, postingsReader, fieldMetadata, blockDecoder);
-    this.nextStringCalculator = new AutomatonNextTermCalculator(compiled);
-    Automaton automaton = Objects.requireNonNull(compiled.automaton);
-    this.runAutomaton = Objects.requireNonNull(compiled.runAutomaton);
-    this.commonSuffixRef = compiled.commonSuffixRef; // maybe null
-    this.commonPrefixRef = Operations.getCommonPrefixBytesRef(automaton); // never null
-
-    this.startTerm = startTerm;
-    assert startTerm == null || StringHelper.startsWith(startTerm, commonPrefixRef);
-    // it is thus also true that startTerm >= commonPrefixRef
-
-    this.seekTerm = startTerm != null ? startTerm : commonPrefixRef;
+    checkIntersectAutomatonType(compiled);
+    automaton = compiled.automaton;
+    runAutomaton = compiled.runAutomaton;
+    finite = compiled.finite;
+    commonSuffix = compiled.commonSuffixRef;
+    minTermLength = getMinTermLength();
+    nextStringCalculator = new AutomatonNextTermCalculator(compiled);
+    seekTerm = startTerm;
   }
 
-  @Override
-  public BytesRef next() throws IOException {
-    clearTermState();
-
-    if (blockHeader == null) { // initial state
-      // note: don't call super.seekCeil here; we have our own logic
-
-      // Set the browser position to the block having the seek term.
-      // Even if -1, it's okay since we'll soon call nextKey().
-      long blockStartFP = getOrCreateDictionaryBrowser().seekBlock(seekTerm);
-      if (isBeyondLastTerm(seekTerm, blockStartFP)) {
-        return null; // EOF
-      }
-
-      // Starting at this block find and load the next matching block.
-      //   note: Since seekBlock was just called, we actually consider the current block as "next".
-      if (nextBlockMatchingPrefix() == false) { // note: starts at seek'ed block, which may have a match
-        return null; // EOF
-      }
+  protected void checkIntersectAutomatonType(CompiledAutomaton compiled) {
+    if (compiled.type != CompiledAutomaton.AUTOMATON_TYPE.NORMAL) {
+      throw new IllegalArgumentException("please use CompiledAutomaton.getTermsEnum instead");
     }
-
-    do {
-
-      // look in the rest of this block.
-      BytesRef term = nextTermInBlockMatching();
-      if (term != null) {
-        return term;
-      }
-
-      // next term dict matching prefix
-    } while (nextBlockMatchingPrefix());
-
-    return null; // EOF
   }
 
   /**
-   * Find the next block that appears to contain terms that could match the automata.
-   * The prefix is the primary clue.  Returns true if at one, or false for no more (EOF).
+   * Computes the minimal length of the terms accepted by the automaton.
+   * This speeds up the term scanning for automatons accepting a finite language.
    */
-  protected boolean nextBlockMatchingPrefix() throws IOException {
-    if (beyondCommonPrefix) {
-      return false; // EOF
+  protected int getMinTermLength() {
```

Comment:
I don't think it will bring value to BlockTree IntersectTermsEnum because by design it already browses the automaton transitions to go to the blocks that have the corresponding prefixes. So this method is probably better here as a specificity of the UniformSplit intersect.
As opposed to common suffix which is generally useful.

Reply:
BlockTree is not the only consumer of automatons.  I've used them in the UnifiedHighligher as well for highlighting automata/MTQs in a place that processes a TokenStream to see which tokens match the query.  Perhaps a quick minimum length check there would be beneficial.  I don't want to suggest any integration work here for you, only the idea that this utility method might have value elsewhere.  On the other hand, perhaps it should not be moved to a general place until there is another use.  
Also, maybe this method ought to have a dedicated test?

Reply:
Good point for the test. I'll try to craft one, when I have some time.

Reply:
Actually a test specific to this method is not easy at all. I don't see a way to randomize and assert the min length without using a code similar to the code itself.
I'll let it as it is today: I leverage the multiple random tests for automatons, regex and fuzzy queries, and I verify that the min term length works (no false negative) with the assertions already present in the nextTermInBlockMatching() method. For each term, we verify with runAutomaton.run() that the acceptance/rejection is correct.